source/1_engine/rex_engine/src/string/stringpool.cpp

```diff
@@ -0,0 +1,88 @@
+#include "rex_engine/string/stringpool.h"
+
+#include "rex_engine/diagnostics/assert.h"
+#include "rex_engine/string/stringentry.h"
+#include "rex_std/functional.h"
+#include "rex_std/internal/utility/pair.h"
+#include "rex_std/unordered_map.h"
+
+namespace rex
+{
+  namespace string_pool
+  {
+    using EntryMap = rsl::unordered_map<StringEntryID, StringEntry>;
+
+    //-------------------------------------------------------------------------
+    EntryMap load_entry_map()
+    {
+      EntryMap map;
+
+      StringEntryID entry_id(0);
+      StringEntry entry("Invalid StringID");
+
+      map.emplace(rsl::move(entry_id), rsl::move(entry));
+
+      return map;
+    }
+
+    //-------------------------------------------------------------------------
+    EntryMap& get_entries()
+    {
+      static EntryMap entries = load_entry_map();
+      return entries;
+    }
+
+    //-------------------------------------------------------------------------
+    const StringEntryID* store(rsl::hash_result hash, rsl::string_view characters)
+    {
+      auto it = get_entries().find(hash);
+      if(it != rsl::cend(get_entries()))
+      {
+        REX_ASSERT_X(rsl::strcmp(characters.data(), it->value.characters().data()) == 0, "Hash collision");
+
+        return &it->key;
+      }
+
+      StringEntryID entry_id(hash);
+      StringEntry entry(characters);
+
+      auto result = get_entries().emplace(rsl::move(entry_id), rsl::move(entry));
+      if(result.emplace_successful)
+      {
+        return &result.inserted_element->key;
+      }
+
+      return nullptr;
```

Comment:
don't you want to error here?
alternatively, in theory I can't think of a way (other than running out of memory) where this branch would get executed.
a hash collision is caught above, the only other reason why it could fail I believe is when it fails to allocate a new entry into the hash map (but even then, it'd return a successful emplace result, it'll likely crash before reaching that point though)

so is it worth returning a pointer and no the element directly?

Reply:
It's there for safety reasons, I could **not** check if the element was inserted successfully but that would be a nasty bug if it would ever fail .. Not sure when but better safe than sorry.

Reply:
would an assert above the return that emplacement failed and returning a default constructed StringEntryID satisfy you in this case or do you prefer the nullptr?
PS: I think it's worth adding an assert regardless

Reply:
It would satisfy me yes. Although I wonder why this pointer is bothering you?

Reply:
it's an indirection

Reply:
Issue has been created for this, it's not causing any issues now.
https://github.com/New-Creta/Rex/issues/105